File: 207-course-schedule/207-course-schedule.cpp
class Solution {
public:
    
    unordered_set<int> uset;
    
    bool hasCycle(vector<vector<int>> &graph,vector<int> &visited,int start=0){
        int n = graph.size();
        uset.insert(start);
        visited[start] = 1;
        bool cycle = false;
        for(int node :graph[start]){
            if(uset.count(node)){
                return true;
            }
            if(!visited[node]){
                cycle = cycle || hasCycle(graph,visited,node);
            }
        }
        
        uset.erase(start);
        
        return cycle;    
        
    }
    
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> graph(numCourses,vector<int>());
        
        for(vector<int> preq : prerequisites){
            graph[preq[0]].push_back(preq[1]);
        }
        
        vector<int> visited(numCourses);
        
        for(int i=0;i<numCourses;i++){
            if(!visited[i]){
                if(hasCycle(graph,visited,i)){
                    return false;
                }
            }
        }
        return true;
        
        

    }
};

File: 207-course-schedule/207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        
        #
        #   bi --> ai
        #
        #   Return true if you can finish all courses.
        #
        #   what is impossible case?
        #          __
        #         /  \-->
        #        0<------1
        #
        #   if closed circle is detected, it is impossible.
        #
        #
        #  Example: It is impossible. Here, we prepare path stack. Start from 0
        #
        #          ---->
        #         /     \
        #        0       1--->2
        #         \     /
        #          <----
        #
        #  Path Stack
        #  [0]
        #  [0,1]
        #  [0,1,2]
        #  [0,1]
        #  [0,1,0]  <--- if coming node is already included in stack, closed circle exists.
        #
        #  if start from 1
        #  Path Stack
        #  [1]
        #  [1,2]
        #  [1]
        #  [1,0]
        #  [1,0,1]  <--- coming node is already included in stack
        #
        #
        #  Example: This is acceptable
        #
        #        4--->
        #       /     \
        #      0       1--->2
        #       \     /
        #        3--->
        #
        #  Path Stack
        #  [0]
        #  [0,4]
        #  [0,4,1]
        #  [0,4,1,2]
        #  [0,4,1]
        #  [0,4]
        #  [0]
        #  [0,3]  <--- ok
        #
        #
        
        def dfs(root) :
            visited[root] = True
            path_stack.append(root)
            for node in graph[root] :
                #print(path_stack,node)
                if node in path_stack : return False
                if not visited[node] :
                    if not dfs(node) :
                        return False
                    path_stack.pop()
            return True
        
        from collections import defaultdict
        from collections import deque
        
        # construct graph
        graph = defaultdict(list)
        for ai,bi in prerequisites :
            graph[ai].append(bi)
            
        # visited record
        visited = [ False ]*numCourses
        
        # for-loop to visit all of nodes
        for initial_course in range(numCourses) :
            if not visited[initial_course] :
                # traverse by dfs
                path_stack = deque()
                if not dfs(initial_course) :  # False -- cycle detected case
                    return False
        return True
